pkg/minikube/update/update.go

```diff
@@ -149,71 +164,66 @@ func getTimeFromFileIfExists(path string) time.Time {
 	return timeInFile
 }
 
-func updateBinary(v semver.Version, downloadBinary, updateLinkPrefix, downloadLinkFormat string) {
-	checksum, err := downloadChecksum(v, downloadBinary, downloadLinkFormat)
-	if err != nil {
-		glog.Errorf("Cannot download checksum: %s", err)
-		os.Exit(1)
-	}
-
-	currentBinary, err := osext.Executable()
-	if err != nil {
-		glog.Errorf("Cannot find current binary to execute: %s", err)
-		os.Exit(1)
-	}
+func downloadAndSaveArchive(v semver.Version, archiveName, downloadLinkFormat string) (string, string, error) {
+	url := fmt.Sprintf(downloadLinkFormat, v, archiveName)
 
-	url := fmt.Sprintf(downloadLinkFormat, v, downloadBinary)
-	updateBinaryFile(url, checksum)
-
-	env := os.Environ()
-	args := os.Args
-	err = syscall.Exec(currentBinary, args, env)
-	if err != nil {
-		glog.Errorf("Failed to execute updated binary %s: %s", currentBinary, err)
-		os.Exit(1)
-	}
-}
-
-func updateBinaryFile(url string, checksum []byte) {
-	fmt.Println("Downloading updated binary")
+	fmt.Println("Downloading updated archive")
 	httpResp, err := http.Get(url)
 	if err != nil {
-		glog.Errorf("Cannot download binary: %s", err)
-		os.Exit(1)
+		return "", "", errors.New("Cannot download archive")
 	}
 	defer func() { _ = httpResp.Body.Close() }()
 
-	binary := httpResp.Body
+	updatedArchive := httpResp.Body
 	if httpResp.ContentLength > 0 {
 		bar := pb.New64(httpResp.ContentLength).SetUnits(pb.U_BYTES)
 		bar.Start()
-		binary = bar.NewProxyReader(binary)
+		updatedArchive = bar.NewProxyReader(updatedArchive)
 		defer func() {
 			<-time.After(bar.RefreshRate)
 			fmt.Println()
 		}()
 	}
-	err = update.Apply(binary, update.Options{
-		Hash:     crypto.SHA256,
-		Checksum: checksum,
-	})
+
+	archiveBytes, err := ioutil.ReadAll(updatedArchive)
 	if err != nil {
-		glog.Errorf("Cannot apply binary update: %s", err)
-		err := update.RollbackError(err)
-		if err != nil {
-			glog.Errorf("Failed to rollback update: %s", err)
+		return "", "", errors.New("Unable to read downloaded archive")
+	}
+
+	// Create a temporary directory inside minishift directory to store archive contents
+	dir, err := ioutil.TempDir("", "download")
+	if err != nil {
+		return "", "", errors.New("Could not create a temporary directory to store archive contents")
+	}
+
+	var downloadedArchive string
+
+	if runtime.GOOS == "windows" {
+		// Store archive in zip file on Windows
+		downloadedArchive = filepath.Join(dir, "minishift.zip")
+		if err := writeArchiveToDisk(downloadedArchive, archiveBytes); err != nil {
+			return "", "", err
+		}
+	} else {
+		// Store archive in tgz file
+		downloadedArchive = filepath.Join(dir, "minishift.tgz")
+
+		if err := writeArchiveToDisk(downloadedArchive, archiveBytes); err != nil {
+			return "", "", err
 		}
-		os.Exit(1)
 	}
+	return downloadedArchive, dir, nil
```

Comment:
Returning the full path so that it can be used in other functions that are going to extract the binary or replace existing binary with updated on.

Agree with `dir` part. Can get rid of it.